--- gcc-7.1.0/gcc/ada/gcc-interface/Makefile.in.orig	2017-09-24 10:40:06.961412300 -0700
+++ gcc-7.1.0/gcc/ada/gcc-interface/Makefile.in	2017-09-24 11:20:30.099303390 -0700
@@ -1238,15 +1238,18 @@
   s-tpopsp.adb<s-tpopsp-posix-foreign.adb \
   $(ATOMICS_TARGET_PAIRS) \
   $(ATOMICS_BUILTINS_TARGET_PAIRS) \
-  system.ads<system-linux-arm.ads
+  system.ads<system-linux-arm.ads \
+  a-exexpr.adb<a-exexpr-gcc.adb \
+  s-excmac.ads<s-excmac-arm.ads
 
   TOOLS_TARGET_PAIRS =  \
     mlib-tgt-specific.adb<mlib-tgt-specific-linux.adb \
     indepsw.adb<indepsw-gnu.adb
 
   EXTRA_GNATRTL_TASKING_OBJS=s-linux.o
-  EXTRA_LIBGNAT_OBJS+=sigtramp-armdroid.o
+  EXTRA_LIBGNAT_OBJS+=raise-gcc.o sigtramp-armdroid.o
   EXTRA_LIBGNAT_SRCS+=sigtramp.h
+  EXTRA_GNATRTL_NONTASKING_OBJS+=g-cppexc.o s-excmac.o
   EH_MECHANISM=-arm
   THREADSLIB =
   GNATLIB_SHARED = gnatlib-shared-dual
@@ -1936,7 +1939,7 @@
 endif
 
 # ARM linux, GNU eabi
-ifeq ($(strip $(filter-out arm% linux-gnueabi%,$(target_cpu) $(target_os))),)
+ifeq ($(strip $(filter-out arm% linux%,$(target_cpu) $(target_os))),)
   LIBGNAT_TARGET_PAIRS = \
   a-intnam.ads<a-intnam-linux.ads \
   s-inmaop.adb<s-inmaop-posix.adb \
@@ -1945,6 +1948,8 @@
   s-osinte.adb<s-osinte-posix.adb \
   s-osinte.ads<s-osinte-linux.ads \
   s-osprim.adb<s-osprim-posix.adb \
+  s-stchop.adb<s-stchop.adb \
+  s-stchop.ads<s-stchop.ads \
   s-taprop.adb<s-taprop-linux.adb \
   s-tasinf.ads<s-tasinf-linux.ads \
   s-tasinf.adb<s-tasinf-linux.adb \
@@ -1956,8 +1961,16 @@
 
   ifeq ($(strip $(filter-out arm%b,$(target_cpu))),)
     EH_MECHANISM=
+    LIBGNAT_TARGET_PAIRS += \
+    system.ads<system-linux-armel.ads
   else
     EH_MECHANISM=-arm
+    LIBGNAT_TARGET_PAIRS += \
+    system.ads<system-linux-armel.ads \
+    a-exexpr.adb<a-exexpr-gcc.adb \
+    s-excmac.ads<s-excmac-arm.ads
+    EXTRA_LIBGNAT_OBJS+=raise-gcc.o
+    EXTRA_GNATRTL_NONTASKING_OBJS+=g-cppexc.o s-excmac.o
   endif
 
   TOOLS_TARGET_PAIRS =  \
@@ -2421,14 +2434,6 @@
   EXTRA_LIBGNAT_OBJS+=raise-gcc.o
   EXTRA_GNATRTL_NONTASKING_OBJS+=g-cppexc.o s-excmac.o
 endif
-
-ifeq ($(EH_MECHANISM),-arm)
-  LIBGNAT_TARGET_PAIRS += \
-    a-exexpr.adb<a-exexpr-gcc.adb \
-    s-excmac.ads<s-excmac-arm.ads
-  EXTRA_LIBGNAT_OBJS+=raise-gcc.o
-  EXTRA_GNATRTL_NONTASKING_OBJS+=g-cppexc.o s-excmac.o
-endif
 
 # Use the Ada 2005 version of Ada.Exceptions by default, unless specified
 # explicitly already. The base files (a-except.ad?) are used only for building
--- gcc-7.1.0/gcc/ada/gcc-interface/Make-lang.in.orig	2017-09-24 10:39:50.994257622 -0700
+++ gcc-7.1.0/gcc/ada/gcc-interface/Make-lang.in	2017-09-24 10:54:07.687155063 -0700
@@ -366,7 +366,8 @@
  ada/s-sopco4.o	\
  ada/s-sopco5.o	\
  ada/s-stache.o	\
- ada/s-stalib.o	\
+ ada/s-stalib.o \
+ ada/s-stchop.o \
  ada/s-stoele.o	\
  ada/s-strcom.o	\
  ada/s-strhas.o	\
@@ -550,6 +551,7 @@
  ada/s-sopco5.o   \
  ada/s-stache.o   \
  ada/s-stalib.o   \
+ ada/s-stchop.o   \
  ada/s-stoele.o   \
  ada/s-strhas.o   \
  ada/s-string.o   \
@@ -969,8 +971,8 @@
    ada/a-uncdea.ads ada/alloc.ads ada/debug.ads ada/hostparm.ads ada/namet.ads \
    ada/opt.ads ada/osint.ads ada/output.ads ada/sdefault.ads ada/sdefault.adb \
    ada/s-exctab.ads ada/s-memory.ads ada/s-os_lib.ads ada/s-parame.ads \
-   ada/s-stalib.ads ada/s-strops.ads ada/s-sopco3.ads ada/s-sopco4.ads \
-   ada/s-sopco5.ads ada/s-string.ads ada/s-traent.ads ada/s-unstyp.ads \
+   ada/s-stalib.ads ada/s-strops.ads ada/s-sopco3.ads ada/s-sopco4.ads ada/s-sopco5.ads \
+   ada/s-stchop.ads ada/s-stchop.adb ada/s-string.ads ada/s-traent.ads ada/s-unstyp.ads \
    ada/s-wchcon.ads ada/system.ads ada/table.adb ada/table.ads ada/tree_io.ads \
    ada/types.ads ada/unchdeal.ads ada/unchconv.ads
 
