--- gcc/Makefile.in.orig	2017-09-25 13:43:06.846411088 -0700
+++ gcc/Makefile.in	2017-09-25 13:43:29.259745355 -0700
@@ -1899,8 +1899,7 @@
 # "nul.s" on Windows. Because on Windows "nul" is a reserved file name.
 # Specify the path to gcc/testsuite/selftests within the srcdir
 # as an argument to -fself-test.
-SELFTEST_FLAGS = -nostdinc -x c /dev/null -S -o /dev/null \
-	-fself-test=$(srcdir)/testsuite/selftests
+SELFTEST_FLAGS = -nostdinc -x c /dev/null -S -o /dev/null
 
 # Run the selftests during the build once we have a driver and a cc1,
 # so that self-test failures are caught as early as possible.
